proposals/lip-Introduce-sparse-Merkle-trees.md

```diff
@@ -0,0 +1,589 @@
+    let keyPrefix2 be the first (h-1) digits of binaryKey2
+    if keyPrefix1 != keyPrefix2:
+        return false
+    let d1 be the digit at position h of binaryKey1
+    let d2 be the digit at position h of binaryKey2
+    return d1==0 and d2==1
+```
+
+```java
+function calculateRoot(siblingHashes, queries):
+    //we add the hash property to each query, used to recompute the tree root
+    for each query q in queries:
+        set q.binaryBitmap to binaryExpansion(q.bitmap) without leading zeros
+        //if value is empty, it is a defaultEmptyNode
+        if q.value is empty bytes:
+            set q.hash to EMPTY_HASH
+        else:
+            let qLeaf=leafNode(q.key,q.value)
+            set q.hash to qLeaf.hash
+
+    let sortedQueries=sortByBitmapAndKey(queries)
+    sortedQueries=filterQueries(sortedQueries)
+    while sortedQueries is non-empty:
+        let q be the first element of sortedQueries
+        //if the binaryBitmap is empty string, we reached the top of the tree
+        if q.binaryBitmap is empty string: 
+            return q.hash
+
+        let b to the first digit of q.binaryBitmap
+        //h equals the height of the node; e.g., the root has h=0      
+        let h=length of q.binaryBitmap
+        let binaryKey=binaryExpansion(q.key)
+
+        //we distinguish three cases for the sibling hash:
+        //1. sibling is next element of sortedQueries
+        let nextQuery = next element of sortedQueries
+        if nextQuery exists and areSiblingQueries(q, nextQuery): 
+            set siblingHash to nextQuery.hash
+            remove nextQuery from sortedQueries
+        //2. sibling is default empty node
+        else if b == 0: 
+            set siblingHash to EMPTY_HASH
+        //3. sibling hash comes from siblingHashes
+        else if b == 1:
+            set siblingHash to the first element of siblingHashes
+            remove the first element of siblingHashes
+
+        let d be the digit at position (h-1) of binaryKey    
+        if d == 0:
+            update q.hash to branchHash(q.hash, siblingHash)
+        else if d == 1:
+            update q.hash to branchHash(siblingHash, q.hash)
+
+        remove the first digit of q.binaryBitmap
+        sortedQueries=sortByBitmapAndKey(sortedQueries)
+        sortedQueries=filterQueries(sortedQueries)
+    
+    throw error
+```
+
+```java
+function verify(queryKeys, proof, merkleRoot):
+    //check that the proof is valid with respect to queryKeys
+    if length of queryKeys != length of proof.queries:
+        return false
+    for each query q in proof.queries:
+        let k be the corresponding key in queryKeys (the element of queryKeys with same index)
+        //q is an inclusion proof for k or a default empty node 
+        if q.key == k:
+            continue
+        //q is an inclusion proof for another leaf node
+        let binaryQueryBitmap = binaryExpansion(q.bitmap) without leading zeros
+        let binaryQueryKey=binaryExpansion(q.key)
```

Comment:
As query key I would understand the key provided in the query. Maybe proof key would be a better name here?

Reply:
Looks good. That is more clear.